HeterogeneousCore/AlpakaCore/interface/alpaka/EDMetadata.h

```diff
@@ -0,0 +1,95 @@
+#ifndef HeterogeneousCore_AlpakaCore_interface_alpaka_EDMetadata_h
+#define HeterogeneousCore_AlpakaCore_interface_alpaka_EDMetadata_h
+
+#include <atomic>
+#include <memory>
+
+#include <alpaka/alpaka.hpp>
+
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/HostOnlyTask.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+  /**
+   * The EDMetadata class provides the exact synchronization
+   * mechanisms for Event data products for backends with asynchronous
+   * Queue. These include
+   * - adding a notification for edm::WaitingTaskWithArenaHolder
+   * - recording an Event
+   * - synchronizing an Event data product and a consuming EDModule
+   *
+   * For synchronous backends the EDMetadata acts as an owner of the
+   * Queue object, as no further synchronization is needed.
+   *
+   * EDMetadata is used as the Metadata class for
+   * edm::DeviceProduct<T>, and is an implementation detail (not
+   * visible to user code).
+   *
+   * TODO: What to do with device-synchronous backends? The data
+   * product needs to be wrapped into the edm::DeviceProduct, but the
+   * EDMetadata class used there does not need anything except "dummy"
+   * implementation of synchronize(). The question is clearly
+   * solvable, so maybe leave it to the time we would actually need
+   * one?
+   */
+
+#ifdef ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED
+  // Host backends with a synchronous queue
+
+  class EDMetadata {
+  public:
+    EDMetadata(std::shared_ptr<Queue> queue) : queue_(std::move(queue)) {}
+
+    // Alpaka operations do not accept a temporary as an argument
+    // TODO: Returning non-const reference here is BAD
+    Queue& queue() const { return *queue_; }
+
+    void recordEvent() {}
+
+  private:
+    std::shared_ptr<Queue> queue_;
```

Comment:
Does this need to be a `shared_ptr` ?
Alpaka queues are already `shared_ptr`s, can they be used by value, or at most wrapped in an `std::optional` ?

Reply:
Ah, I guess it is used to support reusing queue in the `QueueCache`.

Reply:
> Ah, I guess it is used to support reusing queue in the `QueueCache`.

Correct.